PLMS.BLL/ServicesImplementation/TaskService.cs

```diff
@@ -1,206 +1,338 @@
-﻿using PLMS.BLL.DTO;
+﻿using AutoMapper;
+using Microsoft.EntityFrameworkCore;
+using PLMS.BLL.DTO;
+using PLMS.BLL.Filters;
 using PLMS.BLL.ServicesInterfaces;
 using PLMS.DAL.Entities;
-using PLMS.BLL.Filters;
 using PLMS.DAL.Interfaces;
+using System.Linq.Expressions;
 using Task = PLMS.DAL.Entities.Task;
 
 namespace PLMS.BLL.ServicesImplementation
 {
     public class TaskService : ITaskService
     {
-
         private readonly IUnitOfWork _unitOfWork;
 
-        private readonly IRepository<Task> _taskRepository;
+        private readonly IMapper _mapper;
 
-        public TaskService(IUnitOfWork unitOfWork)
+        public TaskService(IUnitOfWork unitOfWork, IMapper mapper)
         {
             _unitOfWork = unitOfWork;
-            _taskRepository = _unitOfWork.GetRepository<Task>();
+            _mapper = mapper;
         }
-        public async System.Threading.Tasks.Task AddTaskAsync(TaskDto taskDto)//ToDo: should be a separate DTO class for adding Task
+
+        public async System.Threading.Tasks.Task AddTaskAsync(AddTaskDto taskDto, string userId)
         {
+            var goalRepository = _unitOfWork.GetRepository<Goal>();
+            var categoryRepository = _unitOfWork.GetRepository<Category>();
+            var priorityRepository = _unitOfWork.GetRepository<Priority>();
+            var statusRepository = _unitOfWork.GetRepository<Status>();
+            var defaultPriority = await priorityRepository.GetByIdAsync(1);
+            var defaultStatus = await statusRepository.GetByIdAsync(1);
+
+            Goal goal;
+            if (taskDto.GoalId != 0)
+            {
+                goal = await goalRepository.GetByIdAsync(taskDto.GoalId);
+                goal ??= await MapToGoalAsync(taskDto, categoryRepository, priorityRepository, statusRepository, defaultPriority, defaultStatus, userId);
+            }
+            else
+            {
+                goal = await MapToGoalAsync(taskDto, categoryRepository, priorityRepository, statusRepository, defaultPriority, defaultStatus, userId);
+            }
```

Comment:
The concern of this method is adding a task, not work with goal and its related entities.
So, utilize the EF navigation properties features and simplify this code

Reply:
Let's review this method later